truffle/src/main/ruby/core/io.rb

```diff
@@ -588,13 +588,12 @@ def self.foreach(name, separator=undefined, limit=undefined, options=undefined)
 
     name = Rubinius::Type.coerce_to_path name
 
+    separator = $/ if undefined.equal?(separator)
```

Comment:
It probably doesn't matter all that much, but this seems to do extra processing that we didn't have before and doesn't seem to make things all that much clearer.